### ISSUE TITLE
CA1508 False positive on string.IsNullOrEmpty

### ISSUE THREAD

**Issue by meziantou** [dotnet/roslyn-analyzers#4549]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508)

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.101](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

```
static void Repro(string value)
{
    string str = "";
    if (value != null)
    {
        str = "abc";
    }

    if (string.IsNullOrEmpty(str)) // CA1508 'string.IsNullOrEmpty(str)' is always 'false'
    {
        // string.IsNullOrEmpty(str) could be true if `value` is null
    }
}
```

If you change the variable initialization to `string str = null;`, the diagnostic is not reported.

It looks similar to #4546. Feel free to close this issue if it is duplicate.

**Comment by Evangelink** [comment]
Are you sure this code snippet does reproduce the issue? If so then it appears to be fixed on master :)

**Comment by mavasani** [comment]
Verified that the issue repros on analyzers from .NET 5.0 SDK as well as `Microsoft.CodeAnalysis.NetAnalyzers` NuGet package `5.0.0`. However, the issue is fixed in master and `Microsoft.CodeAnalysis.NetAnalyzers` NuGet package `5.0.1`. The newer package has not been inserted into .NET 5 SDK, but should be available in [daily builds](https://github.com/dotnet/installer#installers-and-binaries) of .NET 6 SDK or you can directly reference the latest NuGet package itself to get the analyzers with the fix.

Closing the issue.

**Comment by meziantou** [comment]
Yes, I do see the warning when building my project. This is great if it's already fixed :) But just to be sure here's the full repro

````xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
    <NoWarn>$(NoWarn);CA1014</NoWarn>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>
</Project>
````

````c#
namespace ConsoleApp18
{
    class Program
    {
        public static void Repro(string value)
        {
            string str = "";
            if (value != null)
            {
                str = "abc";
            }

            if (string.IsNullOrEmpty(str)) // CA1508 'string.IsNullOrEmpty(str)' is always 'false'
            {
            }
        }
    }
}
````

Right now, I cannot see the warning in VS (it did yesterday), but when building the project using `dotnet build` I get the warning

![image](https://user-images.githubusercontent.com/509220/101802405-631d9a80-3add-11eb-8fcd-6052e1ada429.png)

I'm using a preview version of .NET 5.0. So, here's the output of `dotnet --info`

````
C:\Users\mezia\source\repos\ConsoleApp18\ConsoleApp18>dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.200-preview.20601.7
 Commit:    b3b934bbf2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.200-preview.20601.7\

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

.NET SDKs installed:
  3.1.403 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.200-preview.20601.7 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
````

**Additional notes**

When I remove the namespace, the warning disapear 🤔

**Comment by meziantou** [comment]
> However, the issue is fixed in master and Microsoft.CodeAnalysis.NetAnalyzers NuGet package 5.0.1

Using the package `Microsoft.CodeAnalysis.NetAnalyzers` 5.0.1 works! There is no warning with the code from my previous comment 😊

````
  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.1">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
````

**Comment by mavasani** [comment]
Thanks @meziantou - note that because .NET 5 SDK itself has a much higher bug bar, all fixes to analyzers which are off by default (like CA1508) which were made _after the release_ of initial .NET 5.0 SDK would only be available in .NET 6 SDK or the NuGet package releases of `Microsoft.CodeAnalysis.NetAnalyzers` on nuget.org.

**Comment by meziantou** [comment]
I wasn't aware of that. So, I will reference the NuGet explicitly in my projects to get bug fixes for these useful analyzers!
